### PR TITLE
Get TreeCache OnSyncError return a dedicated error when pump stalls

### PR DESCRIPTION
**Pull request by wayt**
Currently, when the treeCache stalles the error passed to `OnSyncError` hook is `watch channel closed unexpectedly`. Which is missleading.

This PR makes sure we get an appropriate event in such case.

### CODE REVIEW

File: conn.go [hunk 1]
```diff
@@ -927,12 +927,19 @@ func (c *Conn) addWatcher(path string, kind watcherKind, opts watcherOptions) <-
 	key := watcherKey{path, kind}
 	var w watcher
 
+	staleCallback := opts.staleCallback
```
Comment (jpfourny) [inline conn.go:930]: Typo: Rename "stale" to "stall".

File: tree_cache.go [hunk 2]
```diff
@@ -262,11 +262,15 @@ func (tc *TreeCache) Sync(ctx context.Context) (err error) {
 }
 
 func (tc *TreeCache) doSync(ctx context.Context) error {
+	staled := false
```
Comment (jpfourny) [inline tree_cache.go:265]:
```suggestion
	stalled := false
```

File: watcher.go [hunk 3]
```diff
@@ -32,11 +32,19 @@ func WithWatcherReservoirLimit(reservoirLimit int) WatcherOption {
 	}
 }
 
+func WithStaleCallback(staleCallback func()) WatcherOption {
```
Comment (jpfourny) [inline watcher.go:35]:
```suggestion
func WithStallCallback(stallCallback func()) WatcherOption {
```

File: conn.go (hunk 1 above)
Comment (wayt) [reply conn.go:930]: image

File: tree_cache.go [hunk 4]
```diff
@@ -262,11 +265,15 @@ func (tc *TreeCache) Sync(ctx context.Context) (err error) {
...
 func (tc *TreeCache) doSync(ctx context.Context) error {
+	stalled := false
 	// Start a recursive watch, so we do not miss any changes.
 	// We'll catch up with the changes after the initial sync.
 	watchCh, err := tc.conn.AddWatchCtx(ctx, tc.rootPath, true,
 		WithWatcherInvalidateOnDisconnect(),
-		WithWatcherReservoirLimit(tc.reservoirLimit))
+		WithWatcherReservoirLimit(tc.reservoirLimit),
+		WithStallCallback(func() {
+			stalled = true
```
[the remark this answers is not in the document]
Comment (wayt) [reply tree_cache.go:275]: I was sure it was done in the same goroutine, my bad.

[the remark this answers is not in the document]
Comment (jpfourny) [reply tree_cache.go:275]: It will be one of the goroutines in the pump calling this.

**Pull request merged**